crates/sncast/src/lib.rs

```diff
@@ -269,63 +286,92 @@ fn get_account_from_keystore<'a>(
+
+    let variant_info = account_info
+        .get("variant")
+        .expect("Failed to find variant key in the account JSON file");
+    let public_key = parse_to_string(variant_info, "public_key")
+        .ok_or_else(|| anyhow::anyhow!("Failed to get public key from account JSON file"))?;
+    let legacy = variant_info.get("legacy").and_then(Value::as_bool);
+
+    Ok(Account {
+        private_key,
+        public_key,
         address,
-        chain_id,
-        ExecutionEncoding::Legacy,
-    ))
+        salt,
+        deployed,
+        class_hash,
+        legacy,
+    })
 }
 
-fn get_account_from_accounts_file<'a>(
+fn get_account_info_from_accounts_file(
     name: &str,
-    accounts_file_path: &Utf8PathBuf,
-    provider: &'a JsonRpcClient<HttpTransport>,
     chain_id: FieldElement,
-) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
-    account_file_exists(accounts_file_path)?;
-    let account_info = get_account_info(name, chain_id, accounts_file_path)?;
-    let signer = LocalWallet::from(SigningKey::from_secret_scalar(
-        FieldElement::from_hex_be(&account_info.private_key).with_context(|| {
-            format!(
-                "Failed to convert private key = {} to FieldElement",
-                &account_info.private_key
-            )
-        })?,
-    ));
-    let address = FieldElement::from_hex_be(&account_info.address).with_context(|| {
-        format!(
-            "Failed to convert account address = {} to FieldElement",
-            &account_info.address
+    path: &Utf8PathBuf,
+) -> Result<Account> {
+    raise_if_empty(name, "Account name")?;
+    account_file_exists(path)?;
+
+    let file_content =
+        fs::read_to_string(path).with_context(|| format!("Failed to read a file = {path}"))?;
+    let accounts: HashMap<String, HashMap<String, Account>> =
+        serde_json::from_str(&file_content)
+            .with_context(|| format!("Failed to parse file = {path} to JSON"))?;
+    let network_name = chain_id_to_network_name(chain_id);
+    let account = accounts
+        .get(&network_name)
+        .and_then(|accounts_map| accounts_map.get(name))
+        .cloned();
+
+    account.ok_or_else(|| {
+        anyhow!(
+            "Account = {} not found under network = {}",
+            name,
+            network_name
         )
-    })?;
-    let account = SingleOwnerAccount::new(
-        provider,
-        signer,
-        address,
-        chain_id,
-        ExecutionEncoding::Legacy,
-    );
+    })
+}
 
-    Ok(account)
+async fn get_account_encoding(
+    legacy: Option<bool>,
+    class_hash: Option<FieldElement>,
+    address: FieldElement,
+    provider: &JsonRpcClient<HttpTransport>,
+) -> Result<ExecutionEncoding> {
+    if let Some(legacy) = legacy {
+        Ok(map_encoding(legacy))
+    } else {
+        let legacy = check_if_legacy_contract(class_hash, address, provider).await?;
```

Comment:
I'm struggling a bit to understand what problem this function (get_account_encoding) tries to solve, sorry. do you mind sharing a rationale behind this? will we call the network each time to check if an account is legacy, if there is no such info in account file?

<img width="906" alt="obraz" src="https://github.com/foundry-rs/starknet-foundry/assets/9470712/e334e6c5-1630-4ef9-907a-285d22330d82">

Reply:
When we create a `SingleOwnerAccount` we need to provide an encoding parameter that decides how calls will be encoded. There are 2 options: `ExecutionEncoding::Legacy` and `ExecutionEncoding::New`. Until now, we had hardcoded `ExecutionEncoding::Legacy`, but to support Cairo >=1 accounts, we need some logic to choose the correct encoding.

In the account file we can store information for `address`, `class_hash` and `legacy`, but only `address` is required. The function to choose the correct account encoding operates as follows:
- If `legacy` is not `None`, then it determines the encoding solely based on this boolean value. In this case we don't need to execute any RPC calls
- Otherwise it calls `check_if_legacy_contract`, this function first checks the encoding based on `class_hash`, and if `class_hash` is `None`, it checks based on `address`

As I think now the function `check_if_legacy_contract` can be slightly simplified, given that we can always get `ContractClass` only based on `address` using the `getClassAt` starknet method.

Reply:
would it make sense to, if there is no info in the file, do the call to network and store legacy info in the file, so it can be later retrieved? (as opposed to call to network everytime)

Reply:
Generally I think it would make sense. I am not sure though if this should be done in get account logic, it doesn't seem to be an appropriate place in the code to overwrite the accounts file. Currently we make changes in the accounts file in commands `account add/create/deploy`. Logically this makes sense and is expected. I'm leaning towards not implementing logic to store `legacy` data here. If there is any latency issue arise, it's very easy for users to manually add `"legacy": true` to their accounts file.

Reply:
Correct me if I'm wrong, but in script execution we only need to get account once? If this the case then we need to make only one additional RPC call (after making the change that I mentioned in the first message).

Reply:
as discussed yesterday offline, we need to write this info to the file, as there are more requests.

Main idea is to just do it silently in get_account.
The other idea would be to print a warning that legacy info is missing from accounts file, and have something like `account info-update` (name to be determined) that users would be able to call in order to update the file. That way we would by default make eccessive calls, but it would be more fair to users since we will give them an explicit way to update account info and get rid of these calls.

Both approaches have their flaws, but imo both are perfectly acceptable. wdyt?

Reply:
Adding a command to update account information seems reasonable to me. Created an issue to track this https://github.com/foundry-rs/starknet-foundry/issues/1901